src/account/UpgradeableModularAccount.sol

```diff
@@ -445,35 +438,41 @@ contract UpgradeableModularAccount is
         returns (PostExecToRun[] memory postHooksToRun)
     {
         SelectorData storage selectorData = getAccountStorage().selectorData[selector];
-        uint256 preExecHooksLength = selectorData.preHooks.length();
-        uint256 postOnlyHooksLength = selectorData.postOnlyHooks.length();
-        uint256 maxPostExecHooksLength = postOnlyHooksLength;
 
-        // There can only be as many associated post hooks to run as there are pre hooks.
-        for (uint256 i = 0; i < preExecHooksLength; ++i) {
-            (, uint256 count) = selectorData.preHooks.at(i);
-            unchecked {
-                maxPostExecHooksLength += (count + 1);
-            }
+        if (selectorData.denyExecutionCount > 0) {
+            revert AlwaysDenyRule();
         }
 
+        uint256 preExecHooksLength = selectorData.preHooks.length();
+        uint256 postOnlyHooksLength = selectorData.postOnlyHooks.length();
+
         // Overallocate on length - not all of this may get filled up. We set the correct length later.
-        postHooksToRun = new PostExecToRun[](maxPostExecHooksLength);
-        uint256 actualPostHooksToRunLength;
+        postHooksToRun = new PostExecToRun[](preExecHooksLength + postOnlyHooksLength);
+
+        // Copy all post hooks to the array. This happens before any pre hooks are run, so we can
+        // be sure that the set of hooks to run will not be affected by state changes mid-execution.
 
-        // Copy post-only hooks to the array.
+        // Copy post-only hooks.
         for (uint256 i = 0; i < postOnlyHooksLength; ++i) {
-            (bytes32 key,) = selectorData.postOnlyHooks.at(i);
-            postHooksToRun[actualPostHooksToRunLength].postExecHook = _toFunctionReference(key);
-            unchecked {
-                ++actualPostHooksToRunLength;
+            bytes32 key = selectorData.postOnlyHooks.at(i);
+            postHooksToRun[i].postExecHook = _toFunctionReference(key);
+        }
+
+        // Copy associated post hooks to the array.
+        for (uint256 i = 0; i < preExecHooksLength; ++i) {
+            FunctionReference preExecHook = _toFunctionReference(selectorData.preHooks.at(i));
+
+            FunctionReference associatedPostExecHook = selectorData.associatedPostHooks[preExecHook];
+
+            if (associatedPostExecHook.notEq(FunctionReferenceLib._EMPTY_FUNCTION_REFERENCE)) {
```

Comment:
```suggestion
            if (!associatedPostExecHook.isEmpty()) {
```

Reply:
Added a `notEmpty()` function to `FunctionReferenceLib` to make this a bit easier to read, lmk if you prefer the old style.